### ISSUE TITLE
Gherkin: Remove Background from Rule in grammar

### ISSUE THREAD

**Issue by ciaranmcnulty** [cucumber/common#590]
## Summary

Based on conversation at Cukenfest, we will take out Background from Rules and see if valid use cases emerge

## Expected Behavior

This will become a parse error:

```
Feature: My Feature
    Rule: My Rule
        Background: My background
```
## Current Behavior

Currently backgrounds are inlined into the pickles created from the Scenarios

**Comment by aslakhellesoy** [comment]
I had a crack at this (removed `Background?` from the `Feature` rule).

Because of the way Berp works, a rule with a background does *not* cause a parser error. It just generates a `Rule` AST node with description `    Background:\n      Given rb`.

The Berp parser will only attempt to match lines it expects. Everything else becomes a description. I think parser is unable to detect quite a lot of parse errors. /cc @gasparnagy

**Comment by ttutisani** [comment]
Hi. I came here as I was trying to understand whether Gherkin AST produces Rule objects or not. I know that it was introduced as of Gherkin 6, but the AST picture on the start page of this GitHub project does not yet mention Rule in it. So, it is supported yet?

**Comment by gasparnagy** [comment]
@ttutisani It is supported, but the docs have not been updated yet, I guess, but please create a separate issue for this.

**Comment by gasparnagy** [comment]
@aslakhellesoy I see. It was definitely by design like that, but we can question whether this was really a good choice (I am not sure anymore.) I think besides "Background" and maybe "Feature" all the other keywords can be used pretty much everywhere, so the impact of this behavior is not that high. 

If you write "Background" after a "Rule" and start with the first "Given" step, it will anyway cause a synxtax error (if we don't allow backgrounds for rules), because at that point only "Scenario" or another "Rule" can come, so there will be no hidden error caused by that. 

I would go ahead removing the Background for the Rule and maybe have another separate discussion about how the parser should treat the "other" token. We can easily change BERP (if we want) to check every other token and only accept "Other" if no other token is matching. This will not require any change in the language implementations.

**Comment by aslakhellesoy** [comment]
That sounds like a good plan @gasparnagy - good to hear berp can be easily changed.

**Comment by aslakhellesoy** [comment]
IIRC the rationale for not allowing `Background` in `Rule` was that it would provide too much rope for people to hang themselves with. It would be easy to create a mess that's hard to reason about, and we want to avoid that.

**Comment by sealr00t** [comment]
I would suggest to keep `Background` inside `Rule`'s as it provides a very good solution for having separate/multiple `Background` keywords inside one feature file.

**Comment by enkessler** [comment]
Looking at it from a structural/utility perspective:
- Backgrounds are a convenience/clarity mechanism that accomplish this by allowing for less repetition of steps in scenarios. 
- Rules are a convenience/scoping mechanism that accomplish this by demarcating groups of scenarios into organized, related groups.

Their purposes don't seem to conflict at all and their combined effects could still flow in a straightforward and behaviorally understandable manner of  `Before hook execution -> Feature level background execution -> Rule level background execution -> Scenario execution`, even if doing so does increase the overall mechanical complexity of the Cucumber framework.

> IIRC the rationale for not allowing `Background` in `Rule` was that it would provide too much rope for people to hang themselves with. It would be easy to create a mess that's hard to reason about, and we want to avoid that.

While I do contend that it would not be hard to reason about (at least, no more so than any other aspect of Gherkin) unless used in a very bad manner, I can't argue that such very bad usages would not be a likely outcome. That being said, as a user I do like rope and I do like having power and, if some user does end up metaphorically hanging themselves with all of that rope then that is what I would call an 'educational experience'. Call me selfish but I've never been a fan of taking toys away from _everyone_ just because _some_ people don't know how to play nice with them.

**Comment by gasparnagy** [comment]
> Call me selfish but I've never been a fan of taking toys away from everyone just because some people don't know how to play nice with them.

@enkessler I absolutely understand you and I was thinking in the same way before I got  experience with OSS projects that are used by so many people. It was a vivid discussion with many pro and con arguments. Finally we decided this way, with the comment that we might reconsider adding the background support for rules (also maybe tagging) later, once we got more experience about how people use/like the rule structure. We have also somehow realized that making such decisions is better in a live gathering... you should definitely join the next years CukenFest!

**Comment by enkessler** [comment]
@gasparnagy Unfortunately, CukenFest tends to be either geographically or temporally distant to me and so attending such an event necessitates a non-trivial cost in travel expenses, lost work income, or both. Lacking luxuries like vacation days or company sponsorship, I try to minimize event attendance. Even with this past CukenFest being a virtual event, I was not able to attend all of it.

> we might reconsider adding the background support for rules (also maybe tagging) later, once we got more experience about how people use/like the rule structure

The lack of any implementation example of `Rule` is the primary reason that I hadn't yet gotten around to modeling that keyword in [cuke_modeler](https://github.com/enkessler/cuke_modeler), despite its existence since Gherkin 6 or so. With the release of Cucumber 4 (Ruby), however, I can now see it in action and that is what got me to looking into finally modeling the `Rule` keyword and what ultimately led me to this issue thread.

At this point in time, I am able to simultaneously use backgrounds at both the feature level and the rule level and it works exactly as I expected. I plan to use this rope to make a perfectly safe tire swing until such time as a newer release comes along to take away my fun. ;)

**Comment by enkessler** [comment]
*continues to swing on his excess rope enabled tire swing while pondering the implications of having all language platforms' behavior being dictated by the lowest common denominator of language platform aggreement*

**Comment by aslakhellesoy** [comment]
> I had a crack at this (removed `Background?` from the `Rule` grammar rule).
> 
> Because of the way Berp works, a rule with a background does _not_ cause a parser error. It just generates a `Rule` AST node with description ` Background:\n Given rb`.
> 
> The Berp parser will only attempt to match lines it expects. Everything else becomes a description. I think parser is unable to detect quite a lot of parse errors. /cc @gasparnagy

I think the easiest way to implement this is to allow Background after Rule in the parser, but throw an exception in the AST builder.

**Comment by wakaleo** [comment]
Hmmm, I find it quite useful, and use it fairly regularly - if background makes sense for a feature as a whole, why could it not make sense at a rule level, when a rule has several examples associated with it?

**Comment by ciaranmcnulty** [comment]
@wakaleo There's an anti-Background sentiment in some quarters anyhow (see Seb and Gaspar's last book) so extending it to Rules could be seen as making things worse

**Comment by ehuelsmann** [comment]
Hmm. it's my understanding that versions of Cucumber have been released with the `Rule` keyword supporting `Background`, right? If that's the case, then isn't removing `Background` from `Rule` just as invasive as removing `Background` altogether?

I mean to ask: what alternative do we offer those who already invested in either `Background` heavily? Either because they shot themselves in the foot or because they genuinely had a good reason to use the Background abstraction?

**Comment by ehuelsmann** [comment]
> Hmm. it's my understanding that versions of Cucumber have been released with the `Rule` keyword supporting `Background`, right? If that's the case, then isn't removing `Background` from `Rule` just as invasive as removing `Background` altogether?

Ok. So, I just figured out that the `Rule` keyword got the `Background` since its inception (Gherkin v6.0). Gherkin 6 was released in 2018 (or at least the merge did). I'd argue that it's simply too late to remove the background from the Rule keyword. @gasparnagy provides the perfect reason in his original comment explaining why he thinks it shouldn't have been added in the first place:

> with OSS projects that are used by so many people.

**Comment by mpkorstanje** [comment]
We have had backgrounds in rules for a while now.

While we speak of being opinionated, we, its creators clearly thought it was not important to fix the parser. In fact, we even removed the experimental warning from the documentation! If we are opinionated, we are also wishy washy in what our opinion is!

And that is a problem. And we should probably have a discussion about how we got here in the first place. 

But form a practical point of view:

1. Removing background rules will have significant user impact.
2. Backgrounds rules are hidden from the rest of Cucumber through pickles. So the complexity of keeping background rules is contained.
3. Removing background rules will require a deprecation process that spans two major versions of each Cucumber implementation. One to deprecate, one to remove. This will take a significant amount of time and work to coordinate.

So if given the choice between keeping things as is, breaking background rules abruptly or deprecating them I'd say that leaving things as is is the least amount of effort.

> The SpecFlow team doesn't want to implement it

@gasparnagy could you elaborate on the why here. We haven't really discussed the reasons and motivations here.

**Comment by aslakhellesoy** [comment]
There are many good arguments made here for keeping it. Another way to address this is through documentation - make it loud and clear that using backgrounds in rules is *supported*, but potentially a bad idea. I think removing it would be akin to adding a limit of max 1000 lines for a source file - a bit patronising.

[You can't shut the stable door after the horse has bolted](https://i.imgur.com/E1XYHtc.gifv).

**Comment by ciaranmcnulty** [comment]
In terms of the backwards compatibility worries, I know the parser allows them but has there been any user-facing documentation mentioning/encouraging Backgrounds in Rules? If it was never documented as working then arguably it's a bug fix :)

**Comment by davidjgoss** [comment]
@ciaranmcnulty whoops https://github.com/cucumber/docs/pull/440

**Comment by wakaleo** [comment]
> In terms of the backwards compatibility worries, I know the parser allows them but has there been any user-facing documentation mentioning/encouraging Backgrounds in Rules? If it was never documented as working then arguably it's a bug fix :)

I believe I mention it in BDD in Action (2nd Edition) if that counts as user-facing documentation.

**Comment by SabotageAndi** [comment]
> > The SpecFlow team doesn't want to implement it
> 
> @gasparnagy could you elaborate on the why here. We haven't really discussed the reasons and motivations here.

This is nothing that @gasparnagy said @mpkorstanje. That was me and the rest of the SpecFlow team at Tricentis. 
Our motivation was the following: The support for Rules isn't complete in SpecFlow and all our tools. We want to tackle this finally. And so the topic of Background in Rules popped up for us because it is one thing that we don't support yet.
For me the decision we did 2019 (or was is 2018) that we will remove the Background of Rules is still the last decision about this topic and was validated last Thursday again.
So in our case, we also need now to do less to support Rules completely. 
That was our motivation. Only do the work that is really needed.

But if we say, it is now too late to remove the Background. This is also fine for us. Then we will implement it in SpecFlow. We just didn't want to implement something, when we said we will remove it, and is afaik not been implemented in every implementation yet.

We just need now a final decision on what we do with the Background in Rules.

**Comment by aurelien-reeves** [comment]
I share the opinion of @mpkorstanje 

I think that if this is there since the beginning of rules, and it has been at some point documented - even if asked to avoid it - we can not assume that user do not have background for their rules.

And if we remove it, we will certainly have plenty of requests to get it back.

**Comment by luke-hill** [comment]
To put it into perspective, here is an example where multiple backgrounds and rules are useful (Admittedly only one use case).

I also have examples with Backgrounds at the top level and then rule/background. I think since 2018 when we were debating it and 2019 when we accepted it, I've started to find more and more use-cases where it's useful. For me it needs to stay.

If some languages don't want to implement it, I say that's ok. But iirc what @SabotageAndi is saying is true. In 2018 rules were quite new and we were struggling with implementing them. Now they're common place let's implement them wholly >??

```
Feature: Advice feedback

  As a ...
  I want to ....
  So that I ....

  Rule: Rule1
    Background:
      Given I am logged in
      And I am on ....

    Scenario: User can give feedback
      Then I am able to provide feedback

    Scenario: User giving positive feedback
      When I inform ...
      Then I am thanked for providing feedback

    Scenario: User giving detailed free-text negative feedback
      When I inform ....
      And I offer a typed reason for why the advice was not useful
      Then I am thanked for providing feedback

  Rule: Rule2
    Background:
      Given I am on ....

    Scenario: User can give feedback
      Then I am able to provide feedback

    Scenario: User giving positive feedback
      When I inform ...
      Then I am thanked for providing feedback

    Scenario: User giving detailed free-text negative feedback
      When I inform ....
      And I offer a typed reason for why the advice was not useful
      Then I am thanked for providing feedback

    Scenario: User giving detailed negative feedback with a specific reason
      When I inform ...
      And I select a reason that the advice was not useful
      Then I am thanked for providing feedback

    Scenario: User changing their mind about feedback
      When I inform ....
      And I change my mind about giving feedback
      Then I am able to provide feedback
```

**Comment by gasparnagy** [comment]
Yes, it seems to be too late to have a discussion about the usefulness of this as it is already there... It is sad though that we made an agreement by considering pro and con arguments, but finally we (including myself, no blaming here) haven't taken care of applying it and finally the power of status quo have overtaken our arguments.

Now everyone thinks over the usefulness of this feature and certainly will find use cases where it is useful. But (and I am just repeating myself), the question is not whether it is useful for *you*, but whether the complexity and confusion it introduces is overweighted by the benefits for the thousands of the other users. I meet everyday cucumber/specflow users on my courses: 2/3 of them thinks that background executes only once before all scenarios run. These people get a feature file (the ubiquitous language) to review. They will not check the documentation of Gherkin, because they think they understand the background (a nice example of "unknown unknown"). 

As I said, it is too late for the rule backgrounds now (and probably not that important anyway), but I just wanted to highlight that we have to make different considerations when we change Gherkin than any other parts of cucumber. I think I could also find good use cases where *I* could safely use variables, loops or conditional steps in Gherkin, but that is not the point here.

**Comment by ehuelsmann** [comment]
During the weekly update meeting just a minute ago, we confirmed @gasparnagy 's wording followed by the decision to close this issue without further action. Use of backgrounds remains discouraged, but removal is - due to the time it has been available now - no longer an option.
